bug: `Generate Unit Tests` fails for a markdown file on Windows

I've just tried it - it fails with an error 😕 

## Scenario
1. Run `Generate Unit Tests` in a markdown file on Windows

But it does not look like a windows issue. It seems to be a threading problem (pooled thread instead of awt).

```
Access is allowed from event dispatch thread only

java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.lsp4j.jsonrpc.TracingMessageConsumer.consume(TracingMessageConsumer.java:114)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:65)
	... 12 more
Caused by: com.intellij.openapi.diagnostic.RuntimeExceptionWithAttachments: EventQueue.isDispatchThread()=false
Current thread: Thread[pool-5-thread-1,5,main] 1166554505
SystemEventQueueThread: Thread[AWT-EventQueue-0,6,main] 846922080
	at com.intellij.openapi.application.impl.ApplicationImpl.throwThreadAccessException(ApplicationImpl.java:1103)
	at com.intellij.openapi.application.impl.ApplicationImpl.assertIsDispatchThread(ApplicationImpl.java:1086)
	at com.intellij.ide.impl.DataManagerImpl.getDataContext(DataManagerImpl.java:166)
	at com.intellij.ide.impl.DataManagerImpl.getDataContext(DataManagerImpl.java:197)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl.<init>(DialogWrapperPeerImpl.java:85)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerFactoryImpl.createPeer(DialogWrapperPeerFactoryImpl.java:25)
	at com.jetbrains.rdserver.ui.dialogs.BackendDialogWrapperPeerFactory.createPeer(BackendDialogWrapperPeerFactory.kt:51)
	at com.intellij.openapi.ui.DialogWrapper.createPeer(DialogWrapper.java:857)
	at com.intellij.openapi.ui.DialogWrapper.<init>(DialogWrapper.java:232)
	at com.intellij.ui.messages.AlertDialog.<init>(AlertMessagesManager.kt:98)
	at com.intellij.ui.messages.AlertMessagesManager.showMessageDialog(AlertMessagesManager.kt:63)
	at com.intellij.ui.messages.MessagesServiceImpl.showMessageDialog(MessagesServiceImpl.java:49)
	at com.jetbrains.rdserver.ui.BackendMessagesService.showMessageDialog(BackendMessagesService.kt:29)
	at com.intellij.openapi.ui.Messages.showDialog(Messages.java:273)
	at com.intellij.openapi.ui.Messages.showDialog(Messages.java:289)
	at com.intellij.openapi.ui.Messages.showDialog(Messages.java:304)
	at com.intellij.openapi.ui.Messages.showMessageDialog(Messages.java:333)
	at com.intellij.ide.scratch.ScratchRootType.createScratchFile(ScratchRootType.java:95)
	at com.sourcegraph.utils.CodyEditorUtil.createFileOrScratch(CodyEditorUtil.kt:251)
	at com.sourcegraph.cody.agent.CodyAgentService$2.invoke$lambda$9(CodyAgentService.kt:115)
	at com.sourcegraph.cody.agent.CodyAgentClient.openUntitledDocument(CodyAgentClient.java:104)
	... 17 more
```

_Originally posted by @mkondratek in https://github.com/sourcegraph/jetbrains/pull/1525#discussion_r1603862840_

--- a/src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentService.kt
+++ b/src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentService.kt
@@ -107,16 +107,17 @@ class CodyAgentService(private val project: Project) : Disposable {
       }
 
       agent.client.onOpenUntitledDocument = Function { params ->
-        // We always get a file which does not start with file:// there.
-        // They start with untitled:// which is VSCode-specific way
-        // to say that files does not exist on the disk yet.
-        val vf =
-            CodyEditorUtil.createFileOrScratch(project, params.uri, params.content)
-                ?: return@Function false
+        val result = CompletableFuture<Boolean>()
         ApplicationManager.getApplication().invokeAndWait {
+          val vf = CodyEditorUtil.createFileOrScratch(project, params.uri, params.content)
+          if (vf == null) {
+            result.complete(false)
+            return@invokeAndWait
+          }
           CodyEditorUtil.showDocument(project, vf)
+          result.complete(true)
         }
-        true
+        result.get()
       }
 
       agent.client.onRemoteRepoDidChange = Consumer {

--- a/src/main/kotlin/com/sourcegraph/utils/CodyEditorUtil.kt
+++ b/src/main/kotlin/com/sourcegraph/utils/CodyEditorUtil.kt
@@ -24,6 +24,7 @@ import com.intellij.openapi.fileTypes.PlainTextLanguage
 import com.intellij.openapi.project.Project
 import com.intellij.openapi.project.ProjectManager
 import com.intellij.openapi.util.Key
+import com.intellij.openapi.util.SystemInfoRt
 import com.intellij.openapi.util.TextRange
 import com.intellij.openapi.vfs.LocalFileSystem
 import com.intellij.openapi.vfs.VirtualFile
@@ -224,7 +225,12 @@ object CodyEditorUtil {
       return LocalFileSystem.getInstance().refreshAndFindFileByNioFile(fixedUri.toPath())
     } catch (e: URISyntaxException) {
       // Let's try scratch files
-      val fileName = uriString.substringAfterLast('/').substringAfterLast('\\')
+      val fileName =
+          if (SystemInfoRt.isWindows) {
+            uriString.substringAfterLast(':')
+          } else {
+            uriString.substringAfterLast('/').substringAfterLast('\\')
+          }
       return ScratchRootType.getInstance()
           .findFile(project, fileName, ScratchFileService.Option.existing_only)
     }
@@ -244,7 +250,12 @@ object CodyEditorUtil {
       }
       return vf
     } catch (e: URISyntaxException) {
-      val fileName = uriString.substringAfterLast('/').substringAfterLast('\\')
+      val fileName =
+          if (SystemInfoRt.isWindows) {
+            uriString.substringAfterLast(':')
+          } else {
+            uriString.substringAfterLast('/').substringAfterLast('\\')
+          }
       val fileType = FileTypeRegistry.getInstance().getFileTypeByFileName(fileName)
       val language = LanguageUtil.getFileTypeLanguage(fileType) ?: PlainTextLanguage.INSTANCE
       return ScratchRootType.getInstance()

--- a/src/main/kotlin/com/sourcegraph/utils/CodyEditorUtil.kt
+++ b/src/main/kotlin/com/sourcegraph/utils/CodyEditorUtil.kt
@@ -24,7 +24,6 @@ import com.intellij.openapi.fileTypes.PlainTextLanguage
 import com.intellij.openapi.project.Project
 import com.intellij.openapi.project.ProjectManager
 import com.intellij.openapi.util.Key
-import com.intellij.openapi.util.SystemInfoRt
 import com.intellij.openapi.util.TextRange
 import com.intellij.openapi.vfs.LocalFileSystem
 import com.intellij.openapi.vfs.VirtualFile
@@ -225,12 +224,7 @@ object CodyEditorUtil {
       return LocalFileSystem.getInstance().refreshAndFindFileByNioFile(fixedUri.toPath())
     } catch (e: URISyntaxException) {
       // Let's try scratch files
-      val fileName =
-          if (SystemInfoRt.isWindows) {
-            uriString.substringAfterLast(':')
-          } else {
-            uriString.substringAfterLast('/').substringAfterLast('\\')
-          }
+      val fileName = uriString.substringAfterLast(':').trimStart('/', '\\')
       return ScratchRootType.getInstance()
           .findFile(project, fileName, ScratchFileService.Option.existing_only)
     }
@@ -250,12 +244,7 @@ object CodyEditorUtil {
       }
       return vf
     } catch (e: URISyntaxException) {
-      val fileName =
-          if (SystemInfoRt.isWindows) {
-            uriString.substringAfterLast(':')
-          } else {
-            uriString.substringAfterLast('/').substringAfterLast('\\')
-          }
+      val fileName = uriString.substringAfterLast(':').trimStart('/', '\\')
       val fileType = FileTypeRegistry.getInstance().getFileTypeByFileName(fileName)
       val language = LanguageUtil.getFileTypeLanguage(fileType) ?: PlainTextLanguage.INSTANCE
       return ScratchRootType.getInstance()

